### PR TITLE
feat: adding stochastic positional embeddings

### PR DESCRIPTION
**Pull request by kir-7**
Part of #1900 

## Summary
* Adds `add_stochastic_positional_noise` as a utility function to `lightly/models/utils.py`.
* Adds optional use of StoP to: 
      * `lightly/models/modules/ijepa.py` and 
      * `lightly/models/modules/ijepa_timm.py`
* Keeps StoP disabled by default to preserve existing behaviour.
* Adds a new test `tests/utils/test_stochastic_positional_embedding.py` to test the behaviour of the StoP function while disabled and enabled.

## Additional details:
* StoP can be used for masked image modelling (MIM) models and this PR uses it for IJEPA. It can be used for MAE as well, but currently MAEDecoder in `lightly/models/modules`  does not provide functionality to modify the masks directly, these are modified while creating the MAE class (`examples/pytorch/mae.py`). These examples can be modified to evaluate StoP.

### CODE REVIEW

**Comment by kir-7** [comment]
@gabrielfruet  The failing check seems to be regarding test coverage for ijepa.py for from_vit_encoder, should I add tests for this? And is there anything else I need to modify?

**Comment by gabrielfruet** [comment]
@kir-7 Thank you for your contribution! No need to worry, we are planning to remove it #1966 . Either me or Lionel are going to review your PR as soon as we can : )

File: lightly/models/modules/ijepa_timm.py [hunk 1]
```diff
@@ -97,6 +99,9 @@ def __init__(
             ]
         )
 
+        self.use_stop = use_stop
```
Comment (liopeer) [inline lightly/models/modules/ijepa_timm.py]: I saw that the original implementation uses `use_stop`, but imo it's not very descriptive, so I would recommend a more descriptive name and also make sure to update the docstring accordingly. We can also say that we don't use a boolean value at all, but instead only use the `noise_std`, but it's 0 by default.

File: lightly/models/utils.py [hunk 2]
```diff
@@ -1316,3 +1316,46 @@ def apply_masks(x: Tensor, masks: Tensor | list[Tensor]) -> Tensor:
...
+            ``(batch_size, num_tokens, predictor_embed_dim)``.
+        projection:
+            Matrix A used to project gaussian noise to the pos_embedding
+            dimension.
+        noise_dim:
+            Dimension of the sampled gaussian noise before projection.
+        noise_std:
+            Standard deviation of the gaussian noise.
+        enabled:
+            If False, returns ``pos_embeddings`` unchanged.
```
Comment (liopeer) [inline lightly/models/utils.py]: We should use the Google Docstring format and put arguments+description on the same line.

**Review by liopeer — CHANGES_REQUESTED** [review]

File: lightly/models/modules/ijepa_timm.py (hunk 1 above)
Comment (kir-7) [reply lightly/models/modules/ijepa_timm.py]: Appreciate the review. I have removed `use_stop` and now use `noise_std=0.0` by default to disable StoP. I also updated the docstrings accordingly.

File: lightly/models/utils.py (hunk 2 above)
Comment (kir-7) [reply lightly/models/utils.py]: Thanks for the feedback. I updated the docstrings accordingly.

File: lightly/models/utils.py [hunk 3]
```diff
@@ -1316,3 +1316,40 @@ def apply_masks(x: Tensor, masks: Tensor | list[Tensor]) -> Tensor:
...
+
+    noise = torch.normal(
+        mean=0.0,
+        std=noise_std,
+        size=(pos_embeddings.shape[0], pos_embeddings.shape[1], noise_dim),
+        device=pos_embeddings.device,
+        dtype=pos_embeddings.dtype,
+    )
+
+    return pos_embeddings + projection(noise)
```
Comment (liopeer) [inline lightly/models/utils.py]: I actually think we want the projection without the bias term here, i.e. only the matrix multiplication.

**Review by liopeer — COMMENTED** [review]
One last thing to address, otherwise looking good!

File: lightly/models/utils.py (hunk 3 above)
Comment (kir-7) [reply lightly/models/utils.py]: Thanks for the review. The function now directly accepts the projection weight matrix and uses `nn.functional.linear(noise, projection_weight, bias=None)` to project the noise without the bias term.

**Comment by kir-7** [comment]
Hi @liopeer, a quick reminder for this PR. I have finished resolving the previous comments. When you have a chance, could you please take another look and let me know if any further changes are needed? Thanks!

**Review by liopeer — CHANGES_REQUESTED** [review]
It looks mostly correct now, sorry that it took a long time to review. One thing that we need to address though: Adding noise should only happen if the module is in training mode!

**Review by liopeer — APPROVED** [review]
LGTM, thanks a lot!

**Comment by kir-7** [comment]
I also noticed that `IJEPAPredictor.from_vit_encoder` does not properly initial `noise_std`, so model created through this will default `noise_std=0.0`. Should we add an optional `noise_std: float = 0.0` argument to `from_vit_encoder` and pass it to cls(...)?

**Comment by liopeer** [comment]
I think it's fine to leave it as it is, no need to cover this imo.

**Comment by kir-7** [comment]
Sorry for the delay. I merged the latest `master`, resolved the conflicts in `test_ijepa_timm.py`, and updated the tests to cover `noise_std`. CI is currently pending approval to run.
